### DISCUSSION TITLE
Error in startup on Vanilla K8s

### DISCUSSION THREAD

**Discussion by karthik-murugan-hpe** [quay/quay#2728] [Q&A]
Hello,

I'm trying to deploy Quay by installing Quay Operator in my Vanilla K8s. I configured Quay to use an externally managed Postgres and ObjectBucketClaim. 

Quay app's would not boot for reason not clear to me. Could you suggest any clues from the log?
[quay-app.log](https://github.com/quay/quay/files/14507568/quay-app.log)

The host name is just an alias and is not configured in any DNS server. Not sure if this is reason for failure.

**Comment by ralgozino** [comment]
I'm having the same issue, Quay fails to become ready on vanilla K8s deployed using the Quay Operator with routes unmanaged and passing custom certificates and custom config for the storage on a bucket, the rest managed by the Operator.

In the logs I see the same errors as OP, nginx throws an `setpriority(-10) failed (13: Permission denied)` alert but seems to be listening anyway (I can see the readiness probe in the access log), after that message you can see `gunicorn-registry` failing to start and throwing that Assertion error after a line saying:

```
[INFO] [data.logs_model] Using logs model `<data.logs_model.table_logs_model.TableLogsModel object at 0xed6bf4140a30>`.

 Traceback (most recent call last):
   File "src/gevent/_abstract_linkable.py", line 287, in gevent._gevent_c_abstract_linkable.AbstractLinkable._notify_links
   File "src/gevent/_abstract_linkable.py", line 333, in gevent._gevent_c_abstract_linkable.AbstractLinkable._notify_links
 AssertionError: (None, <callback at 0xed6bf4f1a440 args=([],)>)
 <callback at 0xed6bf4f1a440 args=([],)> failed with AssertionError
```

UPDATE: Doing a port-forward I can access Quay's UI and create a user and login. Accessing the health endpoint I can see that both the gunicorn-web and gunicorn-registry are not healthy:

{
  "data": {
    "services": {
      "auth": true,
      "database": true,
      "disk_space": true,
      "registry_gunicorn": false,
      "service_key": true,
      "web_gunicorn": false
    }
  },
  "status_code": 503
}
Setting DEBUGLOG=true does not seem to provide any clues on why the assertions fail.



UPDATE 2: I think I know what is happening.

I set both TLS and Routes as unmanaged, because I want to use an Ingress doing the TLS termination (I don't have Routes available in vanilla K8s).

The Operator still makes me set the ssl.cert and ssl.key in the config bundle otherwise it will complain saying that config is missing.

So Quay gets configured anyway with TLS enabled with the certs that I put in the config bundle (valid, trusted certs BTW).

Here's the issue though:

- Quay listens in HTTP in port 8080 
- Quay listens in HTTPS in port 8443
- The internal health check is performed against the HTTP port (8080)
- The nginx is configured to rewrite the HTTP to HTTPS **but it does not set the right port** so you get a 301 response pointing to HTTPS at port 443 instead of 8443.

You can see this by doing a port forward of both 8080 and 8443 and:
- `curl https://localhost:8443` and you get a valid response
- `curl http://localhost:8080` and you get a redirect to `https://localhost` (without the 8443 port)

```bash
$ curl -I --connect-to <myquaydomain>:8443:localhost:8443 https://<myquaydomain>:8443
HTTP/2 200
server: nginx/1.22.1
date: Thu, 13 Mar 2025 12:04:18 GMT
content-type: text/html; charset=utf-8
content-length: 79050
x-frame-options: DENY
cache-control: no-cache, no-store, must-revalidate
vary: Cookie
set-cookie: _csrf_token=redacted; HttpOnly; Path=/; SameSite=Lax
x-frame-options: DENY
strict-transport-security: max-age=63072000; preload

$ curl -I localhost:8080
HTTP/1.1 301 Moved Permanently
Server: nginx/1.22.1
Date: Thu, 13 Mar 2025 12:06:12 GMT
Content-Type: text/html
Content-Length: 169
Connection: keep-alive
Location: https://localhost/
```


UPDATE 3: FOUND A WORKAROUND

Adding the following settings to the config bundle made the healthchecks work:
```
EXTERNAL_TLS_TERMINATION: false
PREFERRED_URL_SCHEME: https
```

These should be the default values, by if you don't set them the internal healthcheck does not work 🤷 


This is not the ideal configuration though, because Quay is doing the TLS termination, so you have to set the ingress controller to passthrough or tell it that the backend is HTTPS, and manage the certificates at Quay level instead of at the ingress level.

**Comment by ralgozino** [comment]
@karthik-murugan-hpe did you manage to make the Operator Work?
